From 70ec64dc5a2a816d6aa324190a726a85fd749b30 Mon Sep 17 00:00:00 2001
From: Hani Kazmi <hanikazmi@google.com>
Date: Tue, 23 May 2023 17:28:56 +0000
Subject: [PATCH] Update Pip launches to not enter pinned task if in
 background.

Addresses a BAL bypass where Pip could be started without the launcher
being visible.

Bug: 271576718
Test: atest CtsWindowManagerDeviceTestCases:PinnedStackTests
Test: atest android.server.wm.BackgroundActivityLaunchTest#testPipCannotStartFromBackground
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:1848b559059e021d1a923513ca2a936c6212a7ac)
Merged-In: Ibadc9c21f1d23f9904fc11009a9c2a40535db5e0
Change-Id: Ibadc9c21f1d23f9904fc11009a9c2a40535db5e0
---
 services/core/java/com/android/server/wm/ActivityStarter.java | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/wm/ActivityStarter.java b/services/core/java/com/android/server/wm/ActivityStarter.java
index 84cd63424cd1..e6d311f7a9aa 100644
--- a/services/core/java/com/android/server/wm/ActivityStarter.java
+++ b/services/core/java/com/android/server/wm/ActivityStarter.java
@@ -1941,7 +1941,8 @@ class ActivityStarter {
         // If Activity's launching into PiP, move the mStartActivity immediately to pinned mode.
         // Note that mStartActivity and source should be in the same Task at this point.
         if (mOptions != null && mOptions.isLaunchIntoPip()
-                && sourceRecord != null && sourceRecord.getTask() == mStartActivity.getTask()) {
+                && sourceRecord != null && sourceRecord.getTask() == mStartActivity.getTask()
+                && !mRestrictedBgActivity) {
             mRootWindowContainer.moveActivityToPinnedRootTask(mStartActivity,
                     sourceRecord, "launch-into-pip");
         }
-- 
2.41.0.255.g8b1d071c50-goog

